google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -93,15 +133,15 @@ void TracingMessageBatch::Flush() {
     message_spans.swap(message_spans_);
   }
 
-  // TODO(#12528): Handle batches larger than 128.
-  auto batch_sink_parent_span = MakeParentSpan(std::move(message_spans));
+  auto batch_sink_spans = MakeBatchSinkSpans(std::move(message_spans));
 
-  // Set the batch sink as the active span.
+  // Set the batch sink parent span as the active span. This will always be the
+  // first span in the vector.
   auto async_scope = internal::GetTracer(internal::CurrentOptions())
-                         ->WithActiveSpan(batch_sink_parent_span);
+                         ->WithActiveSpan(batch_sink_spans.front());
   {
     std::lock_guard<std::mutex> lk(batch_sink_mu_);
-    batch_sink_spans_.push_back(std::move(batch_sink_parent_span));
+    batch_sink_spans_.swap(batch_sink_spans);
```

Comment:
I'm not following exactly what is going on, so I should probably stay quiet, but this looks strange.  The old `batch_sink_spans_` are discarded?

Reply:
I *think* (@alevenberg may correct me) that there is no "old" `batch_sink_spans_`.  After `Flush()` the expectation is that `FlushCallback()` is called "soon", and that should clear `batch_sink_spans_`.

I think that tells me that (maybe) what we should not even have a member variable.  Maybe `Flush()` should consume a functor that takes the necessary action?  Or maybe it should return a `future<void>` and we attach a callback to it taking whatever action is needed?

Reply:
Maybe return a pair and then take the future attach the callback in the TracingMB and in the BatchingPublisher satisfy the promise?